spatialos-sdk/src/worker/schema.rs

```diff
@@ -0,0 +1,662 @@
+    {
+        let internal = unsafe { Schema_CreateComponentUpdate(C::ID) };
+        let fields = unsafe { Schema_GetComponentUpdateFields(internal) }.into();
+        let mut result = Self { internal, fields };
+
+        // Populate the update fields with the data from `update`.
+        update.into_object(result.fields_mut());
+
+        result
+    }
+
+    pub(crate) fn from_worker_update(
+        internal: *mut Schema_ComponentUpdate,
+    ) -> SchemaComponentUpdate<'a> {
+        let fields = unsafe { Schema_GetComponentUpdateFields(internal) }.into();
+        Self { internal, fields }
+    }
+
+    pub fn id(&self) -> ComponentId {
+        unsafe { Schema_GetComponentUpdateComponentId(self.internal) }
+    }
+
+    fn fields(&self) -> &SchemaObject<'a> {
+        &self.fields
+    }
+
+    fn fields_mut(&mut self) -> &mut SchemaObject<'a> {
+        &mut self.fields
+    }
+}
+
+#[derive(Debug)]
+pub struct SchemaComponentData {
+    pub component_id: ComponentId,
+    pub internal: *mut Schema_ComponentData,
+}
+
+impl SchemaComponentData {
+    pub fn new(component_id: ComponentId) -> SchemaComponentData {
+        SchemaComponentData {
+            component_id,
+            internal: unsafe { Schema_CreateComponentData(component_id) },
+        }
+    }
+
+    pub fn component_id(&self) -> ComponentId {
+        unsafe { Schema_GetComponentDataComponentId(self.internal) }
+    }
+}
+
+#[derive(Debug)]
+pub struct SchemaCommandRequest {
+    pub component_id: ComponentId,
+    pub internal: *mut Schema_CommandRequest,
+}
+
+impl SchemaCommandRequest {
+    pub fn new(component_id: ComponentId, command_index: FieldId) -> SchemaCommandRequest {
+        SchemaCommandRequest {
+            component_id,
+            internal: unsafe { Schema_CreateCommandRequest(component_id, command_index) },
+        }
+    }
+
+    pub fn component_id(&self) -> ComponentId {
+        unsafe { Schema_GetCommandRequestComponentId(self.internal) }
+    }
+
+    pub fn command_index(&self) -> FieldId {
+        unsafe { Schema_GetCommandRequestCommandIndex(self.internal) }
+    }
+}
+
+#[derive(Debug)]
+pub struct SchemaCommandResponse {
+    pub component_id: ComponentId,
+    pub internal: *mut Schema_CommandResponse,
+}
+
+impl SchemaCommandResponse {
+    pub fn new(component_id: u32, command_index: u32) -> SchemaCommandResponse {
+        SchemaCommandResponse {
+            component_id,
+            internal: unsafe { Schema_CreateCommandResponse(component_id, command_index) },
+        }
+    }
+
+    pub fn component_id(&self) -> ComponentId {
+        unsafe { Schema_GetCommandResponseComponentId(self.internal) }
+    }
+
+    pub fn command_index(&self) -> FieldId {
+        unsafe { Schema_GetCommandResponseCommandIndex(self.internal) }
+    }
+}
+
+#[derive(Debug)]
+pub struct SchemaObject<'a> {
+    internal: *mut Schema_Object,
+    _marker: PhantomData<&'a ()>,
+}
+
+impl<'a> SchemaObject<'a> {
```

Comment:
No `index` or `count` methods?

Reply:
Nah, they're not necessary here. The methods on `SchemaObject` are extra sugar to simplify the generated code for the `SchemaObjectType` implementations (they remove the need to bring the specific traits into scope). The process of indexing into fields is always handled by the `SchemaField` impl for the field's type, so it doesn't need to be exposed on `SchemaObject` directly. You can see an example of this in `manual_serialization.rs`.